src/shogun/io/SGIO.h

```diff
@@ -84,124 +87,72 @@ enum EMessageLocation
 #endif
 
 // printf like functions (with additional severity level)
-// for object derived from CSGObject
-#define SG_GCDEBUG(...) {											\
-	if (SG_UNLIKELY(io->loglevel_above(MSG_GCDEBUG)))				\
-		io->message(MSG_GCDEBUG, __PRETTY_FUNCTION__, __FILE__, __LINE__, __VA_ARGS__);	\
-}
+#define SG_IO env()->io()
+#define _SRC_LOC SourceLocation{__FILE__, __LINE__, __PRETTY_FUNCTION__}
 
-#define SG_DEBUG(...) {												\
-	if (SG_UNLIKELY(io->loglevel_above(MSG_DEBUG)))					\
-		io->message(MSG_DEBUG, __PRETTY_FUNCTION__, __FILE__, __LINE__, __VA_ARGS__);	\
+#define SG_GCDEBUG(...) {								\
+	SG_IO->message(MSG_GCDEBUG, _SRC_LOC, __VA_ARGS__);	\
 }
 
-#define SG_OBJ_DEBUG(o,...) {										\
-	if (SG_UNLIKELY(o->io->loglevel_above(MSG_DEBUG)))				\
-		o->io->message(MSG_DEBUG, __PRETTY_FUNCTION__, __FILE__, __LINE__, __VA_ARGS__);	\
+#define SG_DEBUG(...) {									\
+	SG_IO->message(MSG_DEBUG, _SRC_LOC, __VA_ARGS__);	\
 }
 
-
-#define SG_INFO(...) {												\
-	if (SG_UNLIKELY(io->loglevel_above(MSG_INFO)))					\
-		io->message(MSG_INFO, __PRETTY_FUNCTION__, __FILE__, __LINE__, __VA_ARGS__);		\
+#define SG_INFO(...) {									\
+	SG_IO->message(MSG_INFO, _SRC_LOC, __VA_ARGS__);		\
 }
 
-#define SG_CLASS_INFO(c, ...) {										\
-	if (SG_UNLIKELY(c::io->loglevel_above(MSG_INFO)))				\
-		c::io->message(MSG_INFO, __PRETTY_FUNCTION__, __FILE__, __LINE__, __VA_ARGS__);	\
+#define SG_WARNING(...) {							\
+	SG_IO->message(MSG_WARN, _SRC_LOC, __VA_ARGS__);	\
 }
 
-#define SG_WARNING(...) { io->message(MSG_WARN, __PRETTY_FUNCTION__, __FILE__, __LINE__, __VA_ARGS__); }
 #define SG_THROW(ExceptionType, ...)                                           \
 	{                                                                          \
-		io->template error<ExceptionType>(                                     \
-		    MSG_ERROR, __PRETTY_FUNCTION__, __FILE__, __LINE__, __VA_ARGS__);  \
+		SG_IO->template error<ExceptionType>(                                  \
+		    MSG_ERROR, _SRC_LOC, __VA_ARGS__);  \
 	}
 #define SG_ERROR(...) SG_THROW(ShogunException, __VA_ARGS__)
-#define SG_OBJ_ERROR(o, ...)                                                   \
-	{                                                                          \
-		o->io->template error<ShogunException>(                                \
-		    MSG_ERROR, __PRETTY_FUNCTION__, __FILE__, __LINE__, __VA_ARGS__);  \
-	}
-#define SG_CLASS_ERROR(c, ...)                                                 \
-	{                                                                          \
-		c::io->template error<ShogunException>(                                \
-		    MSG_ERROR, __PRETTY_FUNCTION__, __FILE__, __LINE__, __VA_ARGS__);  \
-	}
-#define SG_UNSTABLE(func, ...) { io->message(MSG_WARN, __PRETTY_FUNCTION__, __FILE__, __LINE__, \
+
+#define SG_UNSTABLE(func, ...) { SG_IO->message(MSG_WARN, _SRC_LOC, \
```

Comment:
I never liked this message .... maybe we can remove it and only merge tested/stable code? :D

Reply:
they're mainly used with default constructors for classes that shouldn't support them

Reply:
maybe we can kill them?
@vigsterkr @iglesias @lisitsyn

Reply:
> they're mainly used with default constructors for classes that shouldn't support them

Hmm. Why are they defined?

Reply:
> maybe we can kill them?
> @vigsterkr @iglesias @lisitsyn

It sounds good to me. Just stating the obvious here; let's get rid of the macro uses first and then let's remove the macro.

Reply:
or both at once? :)

Reply:
@iglesias We support default constructors for serialization and deserialization. Also for testing cloning.